### DISCUSSION TITLE
Controlling parallelity / profiling?

### DISCUSSION THREAD

**Discussion by agemuend** [Congyuwang/Rusty-Bitcoin-Explorer#19] [General]
First, thanks a lot for this fantastic project, great work!

One question, I'm currently trying this out on a quad socket Xeon Platinum server with 96 physical cores (4x 24). I/O is done on an Intel P4600 Enterprise NVMe SSD. I use iter_connected_block() with FConnectedBlock. I see that on average it uses only two cores and the SSD is pretty idle as well. I'm wondering where its currently constrained. Are there any knobs I could tune for the threading? Do you maybe have any hints? I'd assume the most work is in the connections / mapping inputs? 

Best Greetings

**Comment by Congyuwang** [comment]
I think perhaps it is using a slower disk for the temporary storage of cache. Perhaps you should try:

bitcoin-explorer = { version = "^1.2", default-features = false }

if you have more than 32GB memory (e.g. 64GB). This configuration uses memory for caching UTXO during processing.

**Reply by agemuend** [reply]
Hey, thanks for the answer! 

I'm already using default-features = false, the machine has around 1,5TB RAM, so no shortage there. 

I think I'll do a firegraph to see where most time is spent.

**Reply by Congyuwang** [reply]
The parallelism is not a very simple parallelism. Because when you are looking up the cache for block N, you got to make sure all transaction outputs from block 1-N has already been updated into the cache.

So, I basically designed a certain crate that allows mixed parallel iteration and sequential iteration. https://github.com/Congyuwang/Synced-Parallel-Iterator
where calling `into_par_iter_sync(fn)` on any iterator executes the iteration in parallel, but produces the output in sequential order. So, we can have the following (pseudo-code):

```rust
(0..heights).into_par_iter_sync(|h| { // here the code are executed in parallel
                            block=read_block(h);
                            update_utxo(block);
                            return block
                   }).into_par_iter_sync(|block| { // here the blocks appear in sequetial order
                            matched_block = match_inputs_to_original_outputs(block); 
                            return matched_block
                   })
```

By default into_par_iter_sync uses an equal number of threads as the number provided by `num_cpus::get()` of `num_cpus` crate.

You might want to check which number does `num_cpus::get()` return on your machine.
